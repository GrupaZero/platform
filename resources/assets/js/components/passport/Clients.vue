<style scoped>
    .action-link {
        cursor: pointer;
    }

    .m-b-none {
        margin-bottom: 0;
    }
</style>

<template>
    <div>
        <div class="card mb-4">
            <div class="card-header">
                <div style="display: flex; justify-content: space-between; align-items: center;">
                    <span>
                       {{ $t('passport.oauth_clients') }}
                    </span>

                    <a class="card-link" @click="showCreateClientForm">
                        {{ $t('passport.create_new_client') }}
                    </a>
                </div>
            </div>

            <div class="card-body">
                <!-- Current Clients -->
                <p class="m-b-none" v-if="clients.length === 0">
                    {{ $t('passport.you_have_not_created_any_oauth_clients') }}
                </p>

                <table class="table table-striped" v-if="clients.length > 0">
                    <thead>
                        <tr>
                            <th>{{ $t('passport.client_id') }}</th>
                            <th>{{ $t('passport.name') }}</th>
                            <th>{{ $t('passport.secret') }}</th>
                            <th></th>
                            <th></th>
                        </tr>
                    </thead>

                    <tbody>
                        <tr v-for="client in clients">
                            <!-- ID -->
                            <td style="vertical-align: middle;">
                                {{ client.id }}
                            </td>

                            <!-- Name -->
                            <td style="vertical-align: middle;">
                                {{ client.name }}
                            </td>

                            <!-- Secret -->
                            <td style="vertical-align: middle;">
                                <code>{{ client.secret }}</code>
                            </td>

                            <!-- Edit Button -->
                            <td style="vertical-align: middle;">
                                <a class="action-link" @click="edit(client)">
                                    {{ $t('passport.edit') }}
                                </a>
                            </td>

                            <!-- Delete Button -->
                            <td style="vertical-align: middle;">
                                <a class="action-link text-danger" @click="destroy(client)">
                                    {{ $t('passport.delete') }}
                                </a>
                            </td>
                        </tr>
                    </tbody>
                </table>
            </div>
        </div>

        <!-- Create Client Modal -->
        <div class="modal fade" id="modal-create-client" tabindex="-1" role="dialog">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <h5 class="modal-title">
                            {{ $t('passport.create_client') }}
                        </h5>
                        <button type="button " class="close" data-dismiss="modal" aria-hidden="true">&times;</button>
                    </div>

                    <div class="modal-body">
                        <!-- Form Errors -->
                        <div class="alert alert-danger" v-if="createForm.errors.length > 0">
                            <p>{{ $t('passport.something_went_wrong') }}</p>
                            <br>
                            <ul>
                                <li v-for="error in createForm.errors">
                                    {{ error }}
                                </li>
                            </ul>
                        </div>

                        <!-- Create Client Form -->
                        <form class="form-horizontal" role="form">
                            <!-- Name -->
                            <div class="form-group">
                                <label class="form-control-label" for="create-client-name">{{ $t('passport.name') }}</label>

                                <input id="create-client-name" type="text" class="form-control"
                                        @keyup.enter="store" v-model="createForm.name">

                                <small class="form-text text-muted">
                                    {{ $t('passport.something_your_users_will_recognize_and_trust') }}
                                </small>
                            </div>

                            <!-- Redirect URL -->
                            <div class="form-group">
                                <label class="form-control-label">{{ $t('passport.redirect_url') }}</label>

                                <input type="text" class="form-control" name="redirect"
                                                @keyup.enter="store" v-model="createForm.redirect">

                                <small class="form-text text-muted">
                                    {{ $t('passport.your_applications_authorization_callback_url') }}
                                </small>
                            </div>
                        </form>
                    </div>

                    <!-- Modal Actions -->
                    <div class="modal-footer">
                        <button type="button" class="btn btn-default" data-dismiss="modal">{{ $t('passport.close') }}</button>

                        <button type="button" class="btn btn-primary" @click="store">
                            {{ $t('passport.create') }}
                        </button>
                    </div>
                </div>
            </div>
        </div>

        <!-- Edit Client Modal -->
        <div class="modal fade" id="modal-edit-client" tabindex="-1" role="dialog">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <button type="button " class="close" data-dismiss="modal" aria-hidden="true">&times;</button>

                        <h4 class="modal-title">
                            {{ $t('passport.edit_client') }}
                        </h4>
                    </div>

                    <div class="modal-body">
                        <!-- Form Errors -->
                        <div class="alert alert-danger" v-if="editForm.errors.length > 0">
                            <p>{{ $t('passport.something_went_wrong') }}</p>
                            <br>
                            <ul>
                                <li v-for="error in editForm.errors">
                                    {{ error }}
                                </li>
                            </ul>
                        </div>

                        <!-- Edit Client Form -->
                        <form class="form-horizontal" role="form">
                            <!-- Name -->
                            <div class="form-group">
                                <label class="col-md-3 control-label">{{ $t('passport.name') }}</label>

                                <div class="col-md-7">
                                    <input id="edit-client-name" type="text" class="form-control"
                                                                @keyup.enter="update" v-model="editForm.name">

                                    <span class="help-block">
                                        {{ $t('passport.something_your_users_will_recognize_and_trust') }}
                                    </span>
                                </div>
                            </div>

                            <!-- Redirect URL -->
                            <div class="form-group">
                                <label class="col-md-3 control-label">{{ $t('passport.redirect_url') }}</label>

                                <div class="col-md-7">
                                    <input type="text" class="form-control" name="redirect"
                                                    @keyup.enter="update" v-model="editForm.redirect">

                                    <span class="help-block">
                                         {{ $t('passport.your_applications_authorization_callback_url') }}
                                    </span>
                                </div>
                            </div>
                        </form>
                    </div>

                    <!-- Modal Actions -->
                    <div class="modal-footer">
                        <button type="button" class="btn btn-default" data-dismiss="modal">{{ $t('passport.close') }}</button>

                        <button type="button" class="btn btn-primary" @click="update">
                            {{ $t('passport.save_changes') }}
                        </button>
                    </div>
                </div>
            </div>
        </div>
    </div>
</template>

<script>
    export default {
        /*
         * The component's data.
         */
        data() {
            return {
                clients: [],

                createForm: {
                    errors: [],
                    name: '',
                    redirect: ''
                },

                editForm: {
                    errors: [],
                    name: '',
                    redirect: ''
                }
            };
        },

        /**
         * Prepare the component (Vue 1.x).
         */
        ready() {
            this.prepareComponent();
        },

        /**
         * Prepare the component (Vue 2.x).
         */
        mounted() {
            this.prepareComponent();
        },

        methods: {
            /**
             * Prepare the component.
             */
            prepareComponent() {
                this.getClients();

                $('#modal-create-client').on('shown.bs.modal', () => {
                    $('#create-client-name').focus();
                });

                $('#modal-edit-client').on('shown.bs.modal', () => {
                    $('#edit-client-name').focus();
                });
            },

            /**
             * Get all of the OAuth clients for the user.
             */
            getClients() {
                axios.get('/oauth/clients')
                        .then(response => {
                            this.clients = response.data;
                        });
            },

            /**
             * Show the form for creating new clients.
             */
            showCreateClientForm() {
                $('#modal-create-client').appendTo("body").modal('show');
            },

            /**
             * Create a new OAuth client for the user.
             */
            store() {
                this.persistClient(
                    'post', '/oauth/clients',
                    this.createForm, '#modal-create-client'
                );
            },

            /**
             * Edit the given client.
             */
            edit(client) {
                this.editForm.id = client.id;
                this.editForm.name = client.name;
                this.editForm.redirect = client.redirect;

                $('#modal-edit-client').appendTo('body').modal('show');
            },

            /**
             * Update the client being edited.
             */
            update() {
                this.persistClient(
                    'put', '/oauth/clients/' + this.editForm.id,
                    this.editForm, '#modal-edit-client'
                );
            },

            /**
             * Persist the client to storage using the given form.
             */
            persistClient(method, uri, form, modal) {
                form.errors = [];

                axios[method](uri, form)
                    .then(response => {
                        this.getClients();

                        form.name = '';
                        form.redirect = '';
                        form.errors = [];

                        $(modal).modal('hide');
                    })
                    .catch(error => {
                        if (typeof error.response.data === 'object') {
                            form.errors = _.flatten(_.toArray(error.response.data));
                        } else {
                            form.errors = [$t('passport.something_went_wrong_please_try_again')];
                        }
                    });
            },

            /**
             * Destroy the given client.
             */
            destroy(client) {
                axios.delete('/oauth/clients/' + client.id)
                        .then(response => {
                            this.getClients();
                        });
            }
        }
    }
</script>
