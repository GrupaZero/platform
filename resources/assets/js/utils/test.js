import Translations from '../lang/locales.js';
import VueI18n from 'vue-i18n';
import {mount} from 'avoriaz';

function vueMount(component, arg) {
    component.i18n = new VueI18n({
        locale: 'en',
        fallbackLocale: 'en',
        messages: Translations
    });
    return mount(component, arg)
}

export {vueMount};