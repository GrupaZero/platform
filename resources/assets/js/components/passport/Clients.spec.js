import test from 'ava'
import {vueMount} from '../../utils/test'
import Bluebird from 'bluebird'
import Clients from './Clients.vue'

/** @TODO Gitlab setup */
/** @TODO Simple tests for other components */
/** @TODO Production build with cache buster ? */

test.beforeEach(() => {
    axiosMock.reset()
    axiosMock.delayResponse = 0
})

test.serial('Init state without clients', t => {
    t.plan(1)

    axiosMock.onGet('/oauth/clients').reply(200, [])

    const wrapper = vueMount(Clients)

    return Bluebird
        .delay()
        .then(() => {
            t.snapshot(wrapper.html())
        })

})

test.serial('Should render OAuth clients table', t => {
    t.plan(3)

    axiosMock.onGet('/oauth/clients')
        .reply(200, [
            {id: 1, name: 'OAuth Client 1'},
            {id: 2, name: 'OAuth Client 2'}
        ])

    const wrapper = vueMount(Clients)
    t.is(0, wrapper.find('tr').length) // Before rendering

    return Bluebird
        .delay()
        .then(() => {
            t.snapshot(wrapper.html())
            t.is(3, wrapper.find('tr').length) // 2 clients + header
        })

})

test.serial('Should render OAuth clients table after 100 ms delay', t => {
    t.plan(3)

    axiosMock.delayResponse = 100

    axiosMock.onGet('/oauth/clients')
        .reply(200, [
            {id: 1, name: 'OAuth Client 1'},
            {id: 2, name: 'OAuth Client 2'}
        ])

    const wrapper = vueMount(Clients)
    t.is(0, wrapper.find('tr').length) // Before rendering

    return Bluebird
        .delay(105)
        .then(() => {
            t.snapshot(wrapper.html())
            t.is(3, wrapper.find('tr').length) // 2 clients + header
        })

})
