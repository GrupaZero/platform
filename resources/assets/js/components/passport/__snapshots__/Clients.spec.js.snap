// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Init state without clients 1`] = `
"<div><div class=\\"panel panel-default\\"><div class=\\"panel-heading\\"><div style=\\"display: flex;\\"><span>
                       OAuth Clients
                    </span><a class=\\"action-link\\">
                        Create New Client
                    </a></div></div><div class=\\"panel-body\\"><p class=\\"m-b-none\\">
                    You have not created any OAuth clients.
                </p><!----></div></div><div id=\\"modal-create-client\\" tabindex=\\"-1\\" role=\\"dialog\\" class=\\"modal fade\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><button type=\\"button \\" data-dismiss=\\"modal\\" aria-hidden=\\"true\\" class=\\"close\\">×</button><h4 class=\\"modal-title\\">
                            Create Client
                        </h4></div><div class=\\"modal-body\\"><!----><form role=\\"form\\" class=\\"form-horizontal\\"><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Name</label><div class=\\"col-md-7\\"><input id=\\"create-client-name\\" type=\\"text\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                        Something your users will recognize and trust.
                                    </span></div></div><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Redirect URL</label><div class=\\"col-md-7\\"><input type=\\"text\\" name=\\"redirect\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                        Your application's authorization callback URL.
                                    </span></div></div></form></div><div class=\\"modal-footer\\"><button type=\\"button\\" data-dismiss=\\"modal\\" class=\\"btn btn-default\\">Close</button><button type=\\"button\\" class=\\"btn btn-primary\\">
                            Create
                        </button></div></div></div></div><div id=\\"modal-edit-client\\" tabindex=\\"-1\\" role=\\"dialog\\" class=\\"modal fade\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><button type=\\"button \\" data-dismiss=\\"modal\\" aria-hidden=\\"true\\" class=\\"close\\">×</button><h4 class=\\"modal-title\\">
                            Edit Client
                        </h4></div><div class=\\"modal-body\\"><!----><form role=\\"form\\" class=\\"form-horizontal\\"><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Name</label><div class=\\"col-md-7\\"><input id=\\"edit-client-name\\" type=\\"text\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                        Something your users will recognize and trust.
                                    </span></div></div><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Redirect URL</label><div class=\\"col-md-7\\"><input type=\\"text\\" name=\\"redirect\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                         Your application's authorization callback URL.
                                    </span></div></div></form></div><div class=\\"modal-footer\\"><button type=\\"button\\" data-dismiss=\\"modal\\" class=\\"btn btn-default\\">Close</button><button type=\\"button\\" class=\\"btn btn-primary\\">
                            Save Changes
                        </button></div></div></div></div></div>"
`;

exports[`Should render OAuth clients table 1`] = `
"<div><div class=\\"panel panel-default\\"><div class=\\"panel-heading\\"><div style=\\"display: flex;\\"><span>
                       OAuth Clients
                    </span><a class=\\"action-link\\">
                        Create New Client
                    </a></div></div><div class=\\"panel-body\\"><!----><table class=\\"table table-borderless m-b-none\\"><thead><tr><th>Client ID</th><th>Name</th><th>Secret</th><th></th><th></th></tr></thead><tbody><tr><td style=\\"vertical-align: middle;\\">
                                1
                            </td><td style=\\"vertical-align: middle;\\">
                                OAuth Client 1
                            </td><td style=\\"vertical-align: middle;\\"><code></code></td><td style=\\"vertical-align: middle;\\"><a class=\\"action-link\\">
                                    Edit
                                </a></td><td style=\\"vertical-align: middle;\\"><a class=\\"action-link text-danger\\">
                                    Delete
                                </a></td></tr><tr><td style=\\"vertical-align: middle;\\">
                                2
                            </td><td style=\\"vertical-align: middle;\\">
                                OAuth Client 2
                            </td><td style=\\"vertical-align: middle;\\"><code></code></td><td style=\\"vertical-align: middle;\\"><a class=\\"action-link\\">
                                    Edit
                                </a></td><td style=\\"vertical-align: middle;\\"><a class=\\"action-link text-danger\\">
                                    Delete
                                </a></td></tr></tbody></table></div></div><div id=\\"modal-create-client\\" tabindex=\\"-1\\" role=\\"dialog\\" class=\\"modal fade\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><button type=\\"button \\" data-dismiss=\\"modal\\" aria-hidden=\\"true\\" class=\\"close\\">×</button><h4 class=\\"modal-title\\">
                            Create Client
                        </h4></div><div class=\\"modal-body\\"><!----><form role=\\"form\\" class=\\"form-horizontal\\"><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Name</label><div class=\\"col-md-7\\"><input id=\\"create-client-name\\" type=\\"text\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                        Something your users will recognize and trust.
                                    </span></div></div><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Redirect URL</label><div class=\\"col-md-7\\"><input type=\\"text\\" name=\\"redirect\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                        Your application's authorization callback URL.
                                    </span></div></div></form></div><div class=\\"modal-footer\\"><button type=\\"button\\" data-dismiss=\\"modal\\" class=\\"btn btn-default\\">Close</button><button type=\\"button\\" class=\\"btn btn-primary\\">
                            Create
                        </button></div></div></div></div><div id=\\"modal-edit-client\\" tabindex=\\"-1\\" role=\\"dialog\\" class=\\"modal fade\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><button type=\\"button \\" data-dismiss=\\"modal\\" aria-hidden=\\"true\\" class=\\"close\\">×</button><h4 class=\\"modal-title\\">
                            Edit Client
                        </h4></div><div class=\\"modal-body\\"><!----><form role=\\"form\\" class=\\"form-horizontal\\"><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Name</label><div class=\\"col-md-7\\"><input id=\\"edit-client-name\\" type=\\"text\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                        Something your users will recognize and trust.
                                    </span></div></div><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Redirect URL</label><div class=\\"col-md-7\\"><input type=\\"text\\" name=\\"redirect\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                         Your application's authorization callback URL.
                                    </span></div></div></form></div><div class=\\"modal-footer\\"><button type=\\"button\\" data-dismiss=\\"modal\\" class=\\"btn btn-default\\">Close</button><button type=\\"button\\" class=\\"btn btn-primary\\">
                            Save Changes
                        </button></div></div></div></div></div>"
`;

exports[`Should render OAuth clients table after 100 ms delay 1`] = `
"<div><div class=\\"panel panel-default\\"><div class=\\"panel-heading\\"><div style=\\"display: flex;\\"><span>
                       OAuth Clients
                    </span><a class=\\"action-link\\">
                        Create New Client
                    </a></div></div><div class=\\"panel-body\\"><!----><table class=\\"table table-borderless m-b-none\\"><thead><tr><th>Client ID</th><th>Name</th><th>Secret</th><th></th><th></th></tr></thead><tbody><tr><td style=\\"vertical-align: middle;\\">
                                1
                            </td><td style=\\"vertical-align: middle;\\">
                                OAuth Client 1
                            </td><td style=\\"vertical-align: middle;\\"><code></code></td><td style=\\"vertical-align: middle;\\"><a class=\\"action-link\\">
                                    Edit
                                </a></td><td style=\\"vertical-align: middle;\\"><a class=\\"action-link text-danger\\">
                                    Delete
                                </a></td></tr><tr><td style=\\"vertical-align: middle;\\">
                                2
                            </td><td style=\\"vertical-align: middle;\\">
                                OAuth Client 2
                            </td><td style=\\"vertical-align: middle;\\"><code></code></td><td style=\\"vertical-align: middle;\\"><a class=\\"action-link\\">
                                    Edit
                                </a></td><td style=\\"vertical-align: middle;\\"><a class=\\"action-link text-danger\\">
                                    Delete
                                </a></td></tr></tbody></table></div></div><div id=\\"modal-create-client\\" tabindex=\\"-1\\" role=\\"dialog\\" class=\\"modal fade\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><button type=\\"button \\" data-dismiss=\\"modal\\" aria-hidden=\\"true\\" class=\\"close\\">×</button><h4 class=\\"modal-title\\">
                            Create Client
                        </h4></div><div class=\\"modal-body\\"><!----><form role=\\"form\\" class=\\"form-horizontal\\"><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Name</label><div class=\\"col-md-7\\"><input id=\\"create-client-name\\" type=\\"text\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                        Something your users will recognize and trust.
                                    </span></div></div><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Redirect URL</label><div class=\\"col-md-7\\"><input type=\\"text\\" name=\\"redirect\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                        Your application's authorization callback URL.
                                    </span></div></div></form></div><div class=\\"modal-footer\\"><button type=\\"button\\" data-dismiss=\\"modal\\" class=\\"btn btn-default\\">Close</button><button type=\\"button\\" class=\\"btn btn-primary\\">
                            Create
                        </button></div></div></div></div><div id=\\"modal-edit-client\\" tabindex=\\"-1\\" role=\\"dialog\\" class=\\"modal fade\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><button type=\\"button \\" data-dismiss=\\"modal\\" aria-hidden=\\"true\\" class=\\"close\\">×</button><h4 class=\\"modal-title\\">
                            Edit Client
                        </h4></div><div class=\\"modal-body\\"><!----><form role=\\"form\\" class=\\"form-horizontal\\"><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Name</label><div class=\\"col-md-7\\"><input id=\\"edit-client-name\\" type=\\"text\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                        Something your users will recognize and trust.
                                    </span></div></div><div class=\\"form-group\\"><label class=\\"col-md-3 control-label\\">Redirect URL</label><div class=\\"col-md-7\\"><input type=\\"text\\" name=\\"redirect\\" class=\\"form-control\\"><span class=\\"help-block\\">
                                         Your application's authorization callback URL.
                                    </span></div></div></form></div><div class=\\"modal-footer\\"><button type=\\"button\\" data-dismiss=\\"modal\\" class=\\"btn btn-default\\">Close</button><button type=\\"button\\" class=\\"btn btn-primary\\">
                            Save Changes
                        </button></div></div></div></div></div>"
`;
